[mono] Decompress ICU data during iOS/tvOS app startup

imhameed opened the pull request:
This change adds support for decompressing ICU data files using the lzfse
decompressor built into Apple OSes.

If the data file path passed to `GlobalizationNative_LoadICUData` does not have
`.lzfse` as a suffix, then the contents of this file will be mapped into
shareable memory and passed directly to ICU.

Otherwise, the contents are decompressed using a fixed-size working buffer and
are stored inside a cache file with a name that contains a decimal encoded
representation of the originating compressed file's inode number, modification
time, and size. This filesystem metadata is extremely likely to change if the
contents of the source file ever changes, so there's no need to compute a
checksum of the data to determine if the cache is still valid. If a cache file
with an appropriate name is already present then it is mapped into shareable
memory and passed to ICU. Stale cache files (defined here to be any file with a
filename that ends with "-icudt.dat.uncompressed" that doesn't exactly match
the desired cache file name) are purged on startup.

`icudt.dat` for mobile is 2126 KiB right now; when compressed with lzfse it
shrinks to 675 KiB. On an iPhone SE 1st gen, this takes 4ms to decompress.

The "decompression framework" also supports lz4, zlib, and lzma. They are all
worse either in decompression time (lzma especially + zlib) or compression
ratio (lz4):

| Algorithm | icudt.dat compressed size | decompression time |
| --------- | ------------------------- | ------------------ |
| lz4       | 1031 KiB                  | 2.41 ms            |
| lzfse     | 675 KiB                   | 4.20 ms            |
| zlib      | 659 KiB                   | 9.61 ms            |
| lzma      | 427 KiB                   | 49.20 ms           |

I am not comfortable adding 50ms to app startup time. On this same
iPhone SE 1st gen, the "Contacts" app takes 166.7ms to display app-generated
pixels. This is end-to-end latency, from the first indication that iOS has
recognized my finger tap, to the display displaying any non-placeholder
content. This was measured with with a 240fps camera. A barebones
xamarin-macios app takes 125ms to display a "Hello world!" message.

This uses the filename's suffix to control decompression for simplicity, but it
would also be possible to instead frame compressed data with a very simple
header consisting of a long-enough magic number and a decompressed payload size
that can be used as a buffer sizing hint to the decompression loop.

Miscellany:
- On iOS, the cache-directory appears to be app-specific. Example: `/var/mobile/Containers/Data/Application/0C22E0D1-26CD-46CB-9EBC-6CF55B513ED1/Library/Caches/`.
- https://www.usenix.org/system/files/conference/osdi14/osdi14-paper-pillai.pdf

imhameed committed:
Strip trailing whitespace.

imhameed committed:
Auto-append a newline to the end of every static icushim error.

imhameed committed:
Add support for on-the-fly decompression of ICU data using Apple-supplied lzfse decompression functions.

am11 commented on src/native/libs/System.Globalization.Native/pal_icushim_static.c:
```diff
@@ -98,60 +104,180 @@ static int32_t load_icu_data(void* pData)
+        size_t path_len = strlen(path);
+        use_lzfse = ends_with(path, path_len, ".lzfse", sizeof(".lzfse") - 1);
+    }
+    #else
+    #endif
     FILE *fp = fopen(path, "rb");
+    if (fp == NULL) {
+    }
```
```suggestion
    #endif
    FILE *fp = fopen(path, "rb");
    if (fp == NULL)
    {
        goto error;
    }
```

am11 commented on src/native/libs/System.Globalization.Native/pal_icushim_static.c:
```diff
@@ -98,60 +104,180 @@ static int32_t load_icu_data(void* pData)
+    fp = NULL;
+
+    #if defined(USE_APPLE_DECOMPRESSION)
+    if (use_lzfse) {
```
Maybe just use strcmp and delete use_lzfse and ends_with:
```suggestion
    if (strcmp(strrchr(path, '.'), ".lzfse") == 0)
```

am11 commented on src/native/libs/System.Globalization.Native/pal_icushim_static.c:
```diff
@@ -98,60 +104,180 @@ static int32_t load_icu_data(void* pData)
+        file_buf = NULL;
+    }
+    #endif
+    if (icu_data == NULL) {
```
nit: `{` on its own line (here and other places).

imhameed committed:
Address (some) feedback

am11 commented:
CI error was (a warning as error) about `use_lzfse`  being unused variable. We can move strcmp statement down to the only place where it is used, and delete `use_lzfse`.

imhameed committed:
Match the brace style used by other files in native/libs/System.Globalization.Native

rolfbjarne reviewed (commented):
I have a couple of thoughts:

* This allocates a big chunk of memory (the size of the entire ICU, ~2 MB).
* It happens on every app launch.

The first one is a problem in certain cases, because the max memory limit is _very_ low for some app extensions (10mb), and in the past we've used about half of that, leaving the other half for the customers. This will take away 40% of the customer's memory budget in a single allocation.

Suggestion:

Would it be possible to save the decompressed output to a (somewhat temporary) file, then mmap that file in read-only mode? That will make the memory purgeable by the OS, and it won't count towards whatever memory limit the OS enforces.

If you only do this if the file doesn't already exist (matching by timestamp, hashcode, or something else), then you'll also solve the second issue.

imhameed committed:
Fix silly typo

imhameed committed:
Remove some detritus from early exploration

steveisok commented:
@imhameed I think you're going to need to add compression to https://github.com/imhameed/runtime/blob/e516581e282a23fc8340475d8c9860d4c787b5e7/src/tasks/AppleAppBuilder/Templates/CMakeLists.txt.template#L59-L63 in order to get the tests building.

imhameed committed:
Map possibly-compressed ICU data files into shareable, purgeable memory.

Decompress ICU data in fixed memory. Cache files are identified with
by the source file's inode number, size, and second-truncated
modification time.

imhameed committed:
Link against the Compression framework in AppleAppBuilder

rolfbjarne commented on src/native/libs/System.Globalization.Native/pal_icushim_static.c:
```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
+            if (bytes_written == -1)
+            {
+                if (last_error == EINTR) continue;
+                log_shim_error("apple_decompress_to_fd: Error during write().");
```
There's no `goto error;` here?

rolfbjarne commented on src/native/libs/System.Globalization.Native/pal_icushim_static.c:
```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
+            }
+            else if (bytes_written == 0)
+            {
+                log_shim_error("apple_decompress_to_fd: write() returned 0.");
```
Nor here?

rolfbjarne commented on src/native/libs/System.Globalization.Native/pal_icushim_static.c:
```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
+
+    if (icu_data == NULL)
+    {
+        log_shim_error("Failed to load ICU data.");
```
Is this supposed to fall through and call `load_icu_data` even if there's no icu_data?

imhameed replied:
No. This was an erroneous omission.

rolfbjarne commented on src/native/libs/System.Globalization.Native/pal_icushim_static.c:
```diff
@@ -98,57 +115,332 @@ static int32_t load_icu_data(void* pData)
+            {
+                log_shim_error("apple_mmap_icu_data: failed to rename %s to %s", tmp_file, cache_file);
+                goto error;
+            }
```
It doesn't look like you're deleting old files here; maybe iterate over all other "icudt-*.dat.decompressed" files and delete them?

imhameed committed:
Remove stale cache files

rolfbjarne commented on src/native/libs/System.Globalization.Native/pal_icushim_static.c:
```diff
@@ -98,57 +116,356 @@ static int32_t load_icu_data(void* pData)
+            if (bytes_written == -1)
+            {
+                if (last_error == EINTR) continue;
+                log_shim_error("apple_decompress_to_fd: Error during write().");
```
What will the app do if the write fails (or anything else here fails)? Will it exit/crash, or will it fall back to executing, but without any ICU available (whatever that means)?

One reasonable cause for failure is the device running out of disk space.

steveisok replied:
If ICU fails to load, the app will terminate with a call to `Environment::FailFast`

imhameed replied:
> What will the app do if the write fails (or anything else here fails)? Will it exit/crash, or will it fall back to executing, but without any ICU available (whatever that means)?
> 
> One reasonable cause for failure is the device running out of disk space.

From testing: the app will continue without ICU data available, so internationalization-related functions will return no data or will fail, exactly as if icudt.dat is not available. We could try decompressing into private memory if writing fails, and only fall back to this degraded failure mode if both disk and memory are unavailable.

tarekgh commented:
> Hmm - that's odd. I would have expected the app to exit. I don't think this is a mode that we should fall back into.

I agree with @steveisok. The behavior should be deterministic and not surprising to the users. In genral, we always fail fast if anything goes wrong. This gives the user the user to know what is wrong and fix it instead of having the user experience multiple failures during the runtime. We'll need to have a failure message be clear though.

SamMonoRT commented:
@steveisok @tarekgh  @imhameed  - Given all the testing done on the PR, and the fact this is not going to change any functionality till we make the change in xamain macios to use this, we should create a separate issue for the FastFail case and merge this in the current state.

steveisok reviewed (approved):
I'm fine with a follow up on the behavior.
